### PR TITLE
Docker 4

### PR DESCRIPTION
**Pull request by ivanov2103**
# Выполнено ДЗ

- [x] Основное ДЗ
- [x] Задание со *

## В процессе сделано:
- Viewed how working network drivers: *none*, *host*.  
Question (driver *host*): "Каков результат? Что выдал docker ps? Как думаете почему?"  
Answer: "Запустился только один контейнер, потому что несколько сервисов с одним и тем же портом не могут быть запущены."  
### **\***  
Containers with driver *host* work in host network namespace, host *default* network namespace used by this containers.  
  
**docker ps | grep -v "CONTAINER ID" ; echo "===namespaces===" ; ip netns**  
*===namespaces===  
default*  
**docker run --network host -d nginx**  
*dc3f3b9580f80f91f7b16504a769d3b59003adad64e3552313bc8c88e662ff6b*  
**docker ps | grep -v "CONTAINER ID" ; echo "===namespaces===" ; ip netns**  
*dc3f3b9580f8        nginx               "nginx -g 'daemon of…"   3 seconds ago       Up 3 seconds                            confident_morse  
===namespaces===  
default*

Container with driver *none* working in dedicated network namespace with localhost interface only.  
  
**docker run --network none -d nginx**  
*d496df048c14c7f3f460a4e723bbcb3204371a7f0a7f69c260957661a4da188c*  
**docker ps | grep -v "CONTAINER ID" ; echo "===namespaces===" ; ip netns**  
*d496df048c14        nginx               "nginx -g 'daemon of…"   5 seconds ago       Up 4 seconds                            competent_bhabha  
dc3f3b9580f8        nginx               "nginx -g 'daemon of…"   3 minutes ago       Up 3 minutes                            confident_morse  
===namespaces===  
2cb70fb43f1c  
default*  
**ip netns exec 2cb70fb43f1c ip -s link**  
*1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN mode DEFAULT group default qlen 1000  
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00  
    RX: bytes  packets  errors  dropped overrun mcast  
    0          0        0       0       0       0  
    TX: bytes  packets  errors  dropped carrier collsns  
    0          0        0       0       0       0*
    
- Viewed how working network driver *bridge*.  
Launched containers in one separate bridge networks with *--network-alias* for name resolution.  
Launched containers in several dedicated bridge networks with *--name* for name resolution and connected "post" and "comment" containers to bridge "ui" container.  
### **\***  
Bridge networks docker host and bridge interfaces of containers was viewed.  
Viewed iptables rules for containers access to external network (MASQUERADE source address translation for all bridge networks), DNAT rule for translation all input connections on 9292 port to "ui" container.  
For listeting 9292 port running docker-proxy process.  

- Installed docker-compose, created *docker-compose.yml* declaration and *.env* file with some enviroment variables for docker-compose.  
Adapted docker-compose.yml for launching containers in several dedicated bridge networks.  
### **\***  

Default project name may be change by *"-p"* command option, by the *"COMPOSE\_PROJECT\_NAME"* environment and *"project_name"* instruction in .yml file.  
### **\***  

I wasn't found how coping files to container by docker-compose and was used volumes instruction for mount directory with app code to container: 
[https://stackoverflow.com/questions/42877801/how-to-sync-code-between-container-and-host-using-docker-compose](https://stackoverflow.com/questions/42877801/how-to-sync-code-between-container-and-host-using-docker-compose)  
[https://docs.docker.com/compose/extends/#understand-the-extends-configuration](https://docs.docker.com/compose/extends/#understand-the-extends-configuration)  

Src files was copied to docker-host by rsync:  
**rsync -avzh ~/ivanov2103_microservices/src/ui ~/ivanov2103_microservices/src/comment ~/ivanov2103_microservices/src/post-py appuser@docker-host:/home/appuser/src**  

Checking volumes:  
**docker exec comment ls -l /app/TestFile ; ssh appuser@35.195.242.58 touch /home/appuser/src/comment/TestFile; docker exec comment ls -l /app/TestFile**  
*ls: /app/TestFile: No such file or directory  
Warning: Permanently added '35.195.242.58' (ECDSA) to the list of known hosts.  
-rw-rw-r--    1 1002     1003             0 Jun 23 07:12 /app/TestFile*  

Checking debug and number of workers (finded in log):  
**docker logs -f comment**  
*...  
[1] \* Process workers: 2  
...  
D, [2018-06-23T07:03:27.748739 #8] DEBUG -- : MONGODB | Server comment_db:27017 initializing.*  
## Как запустить проект:
- Run **docker-compose up -d**
## Как проверить работоспособность:
- Reddit site with comments should be accessible by url http://docker-host:9292
## PR checklisk
- [x] Выставил label с номером домашнего задания
- [x] Выставил label с темой домашнего задания

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -3,192 +3,100 @@ ivanov2103 Infra repository
...
-Moved DB role to other git repository and plugged it to "infra" repository. Checked this by creating terraform stage envinroment and playing site.yml - was succesful.  
-Added checking my DB role to TravisCI (used example):https://github.com/ivanov2103/ansible_role_dbserver - build status badge in title README.md.  
-Added build notification to my Slack channel \#igor_ivanov.  
-In integration process I had some errors:  
-*"GoogleAuthError('PyCrypto library required for '\nlibcloud.common.google.GoogleAuthError: 'PyCrypto library required for Service Account Authentication.'\n"}*  
-*ERROR: Ansible version '2.3.0.0' not supported.  Molecule only supports Ansible versions '>=2.5' with Python version '(3, 6)'*  
-For correcting this errors was changed install instruction in .travis.yml: "- **travis_wait** pip install **ansible>=2.5** molecule apache-libcloud **pycrypto**"  
 
 ## Homework-13  
-- I was studed create and basic manage containers (run, start/stop/kill, view, delete), running process in to container, creating image from container.
```
Comment (Nklya) [inline README.md:133]: Вот здесь начинается работа в microservices репе

File: README.md [hunk 2]
```diff
@@ -3,192 +3,100 @@ ivanov2103 Infra repository
...
-Created ansible playbooks for install Docker and launching container with applications (https://docs.ansible.com/ansible/latest/modules/docker_container_module.html).  
-To save time made a simplified configurations from previous homeworks.  
-Was checked access reddit by url http://IP_instancess:9292 - passed.  
 
 ## Homework-15  
-- Application was divided to three components. For each of this was created image.   
-Question: "Обратите внимание! Cборка ui началась не с первого шага. Подумайте - почему?"  
-Answer: "Перед сборкой образа ui был собран образ comment с идентичными слоями, которые были переиспользованы."  
-Launched our containers in dedicated network with network aliases.  
+
```
Comment (Nklya) [inline README.md:26]: И в чем состоит смысл удалить описание того, что было сделано в прошлых ДЗ???
Было бы еще понятно, если бы остались только те ДЗ, которые сделаны в этом репозитории.
Но это тоже не так.

File: README.md [hunk 3]
```diff
@@ -3,192 +3,100 @@ ivanov2103 Infra repository
...
+### **\***  
+Bridge networks docker host and bridge interfaces of containers was viewed.  
+Viewed iptables rules for containers access to external network (MASQUERADE source address translation for all bridge networks), DNAT rule for translation all input connections on 9292 port to "ui" container.  
+For listeting 9292 port running docker-proxy process.  
+
+- Installed docker-compose, created *docker-compose.yml* declaration and *.env* file with some enviroment variables for docker-compose.  
+Adapted docker-compose.yml for launching containers in several dedicated bridge networks.  
+### **\***  
+
+Default project name may be change by *"-p"* command option, by the *"COMPOSE\_PROJECT\_NAME"* environment and *"project_name"* instruction in .yml file.
```
Comment (Nklya) [inline README.md:80]: 👍

**Review by Nklya — APPROVED** [review]
Добрый вечер.

Игорь, условия ДЗ выполнены, все выглядит относительно хорошо 👍

File: README.md (hunk 2 above)
Comment (ivanov2103) [reply README.md:26]: А в чем суть поиска смысла здесь? Это усложняет проверку Д.З.?

**Pull request merged**